File: 0045-jump-game-ii/0045-jump-game-ii.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
    int n = nums.size();
    if (n == 1) return 0;
    
    int jumps = 0;
    int currentEnd = 0;
    int farthest = 0;

    for (int i = 0; i < n - 1; ++i) {
        farthest = max(farthest, i + nums[i]);
        if (i == currentEnd) {
            jumps++;
            currentEnd = farthest;
            if (currentEnd >= n - 1) break;
        }
    }

    return jumps;
    }
};

File: 0045-jump-game-ii/0045-jump-game-ii.rs
impl Solution {
    pub fn jump(nums: Vec<i32>) -> i32 {
        if nums.len() == 1 {
            return 0;
        }
        let mut jump = 0;
        let mut current_end = 0;
        let mut farthest = 0;

        for i in 0..nums.len() {
            let temp = nums[i] as usize + i;
            if temp > farthest {
                farthest = temp;
            }
            if i == current_end {
                current_end = farthest;
                jump += 1;
            }
            if current_end >= nums.len() - 1 {
                break;
            }
        }
        jump
    }
}